README.md
# Cellular automata

Image or text rendering of {2→9}-states cellular automata.

## [Use here](https://eliseduverdier.fr/cellular_automata/)

- ` GET /cellular_automata/img.php ? s=2 & ...`
- ` GET /cellular_automata/text.php ? s=8 & ...`

| parameter name | shortcut | defines                                                                                                                |
| -------------- | -------- | ---------------------------------------------------------------------------------------------------------------------- |
| states         | s        | The number of states (2 to 9)                                                                                          |
| order          | o        | The order (`1` or `2`)                                                                                                 |
| width          | w        | The width in pixels                                                                                                    |
| height         | h        | The height in pixels                                                                                                   |
| pixel_size     | p        | A cell size in pixels                                                                                                  |
| rule_number    | n        | The rule number or `random` ( `{0 → 256}` for 2 states, `{0 → 134217728}` for 3, etc) (try `110` or `73` for 2 states) |
| random_start   | start    | The first line is random or a single centered point                                                                    |
| color0         | bg       | The base color                                                                                                         |
| color1         | c1       | The first color                                                                                                        |
| color2         | c2       | The second color                                                                                                       |
| color3         | c3       | The third color, etc                                                                                                   |

All parameters are optional, default are defined for everything

### The App

<img src="_design/screenshots/v4.png" width="400" title="Generator for the four–states automata">

## How to use locally

From the folder, launch `php -S localhost:1234` and go to that url. Needs PHP8 (WIP: will run on docker later)

### Run tests

`php app/tests.php`

## The code

That was my first personal project as a brand new developer :) The code was therefore very minimalist (if not plain dirty).

I'm occasionaly trying to keep it in shape and adding new features. _The only rule is no framework, no external library, as much from scratch as possible_.

## TODO

- [x] Add more than 2 states
- [x] Make a text automata
- [x] Add 2nd order
- [ ] Add docker
- [x] Add tests

## Last changes ([see all](CHANGELOG.mg))

### [4.2] - 2021-08-23

- New UI

# Images

### First order

**2 states**

<img src="_design/automata/order1/s2-Rule101.png" height=200 title="Rule 110"> <img src="_design/automata/order1/s2-Rule110.png" height=200 title="Rule 110"> <img src="_design/automata/order1/s2-Rule18.png" height=200 title="Rule 18"> <img src="_design/automata/order1/s2-Rule210.png" height=200 title="Rule 210"> <img src="_design/automata/order1/s2-Rule73.png" height=200 title="Rule 73">

**3 states**

<img src="_design/automata/order1/s3-Rule2059a.png" height=200 title="Rule 2059a"> <img src="_design/automata/order1/s3-Rule2059b.png" height=200 title="Rule 2059b"> <img src="_design/automata/order1/s3-Rule2059.png" height=200 title="Rule 2059"> <img src="_design/automata/order1/s3-Rule4480.png" height=200 title="Rule 4480"> <img src="_design/automata/order1/s3-Rule5404.png" height=200 title="Rule 5404"> <img src="_design/automata/order1/s3-Rule5465.png" height=200 title="Rule 5465"> <img src="_design/automata/order1/s3-Rule910.png" height=200 title="Rule 910">

**5 states**

<img src="_design/automata/order1/s5-Rule161727.png" height=200 title="Rule 161727">

### Second order

**2 states**

<img src="_design/automata/order2/Rule150.png" height=200 title="Rule 150"> <img src="_design/automata/order2/Rule18.png" height=200 title="Rule 18"> <img src="_design/automata/order2/Rule227.png" height=200 title="Rule 227"> <img src="_design/automata/order2/Rule73.png" height=200 title="Rule 73">

**5 states**

<img src="_design/automata/order2/s5-Rule119545276.png" height=200 title="Rule 119545276"> <img src="_design/automata/order2/s5-Rule66133019.png" height=200 title="Rule 66133019">

## Bonus

```
◌◌✻✻◉◌◌✻◉✻◌◌◌◌✻◌◉◌◉◉◌✻✻✻◉◌✻✻◌✻◌✻◉◉◉◉✻◌◌◌✻◉◌◉◌◌◌◌◌◉◉✻◌◌✻◉✻✻◉✻◌◌◌◌✻✻✻
◉◉◉◉◉◌◉◉◉◌◉◉◉◉◉◉◉◌◉✻◉◉◉◉◉◉◉◉◉◉◉◉◉◌◌◉◌◉◉◉◉◉◌◉◌◉◉◉◉◉◉◌◉◉◉◉◌◉◉◌◉◉◉◉◉◉◉
◌◌◌◌✻◌◉◌✻◌◉◌◌◌◌◌✻◌◉◉◉◌◌◌◌◌◌◌◌◌◌◌✻◌◉◉◌◉◌◌◌✻◌◉◌◉◌◌◌◌✻◌◉◌◌✻◌◉✻◌◉◌◌◌◌◌◌
◉◉◉◉◉◉◉◉◉◉◉◌◉◉◉◉◉◉◉◌✻◌◉◉◉◉◉◉◉◉◉◉◉◉◉✻◌◉◌◉◉◉◉◉◌◉◌◉◉◉◉◉◉◌◉◉◉◉◌◉◉◌◉◉◉◉◉
◌◌◌◌◌◌◌◌◌◌✻◌◉◌◌◌◌◌✻◉◉◉◉◌◌◌◌◌◌◌◌◌◌◌◉◌◉◉◌◉◌◌◌✻◌◉◌◉◌◌◌◌✻◌◉◌◌✻◌◉✻◌◉◌◌◌◌
◉◉◉◉◉◉◉◉◉◉◉◉◉◌◉◉◉◉◉◉◌◌✻◌◉◉◉◉◉◉◉◉◉◉◉◌◉✻◌◉◌◉◉◉◉◉◌◉◌◉◉◉◉◉◉◌◉◉◉◉◌◉◉◌◉◉◉
◌◌◌◌◌◌◌◌◌◌◌◌✻◌◉◌◌◌◌✻◌◉◉◉◉◌◌◌◌◌◌◌◌◌✻◌◉◌◉◉◌◉◌◌◌✻◌◉◌◉◌◌◌◌✻◌◉◌◌✻◌◉✻◌◉◌◌
◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◌◉◉◉◉◉◉◌◌✻◌◉◉◉◉◉◉◉◉◉◉◉◌◉✻◌◉◌◉◉◉◉◉◌◉◌◉◉◉◉◉◉◌◉◉◉◉◌◉◉◌◉
◌◌◌◌◌◌◌◌◌◌◌◌◌◌✻◌◉◌◌◌◌✻◌◉◉◉◉◌◌◌◌◌◌◌◌◌✻◌◉◌◉◉◌◉◌◌◌✻◌◉◌◉◌◌◌◌✻◌◉◌◌✻◌◉✻◌◉
◌◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◌◉◉◉◉◉◉◌◌✻◌◉◉◉◉◉◉◉◉◉◉◉◌◉✻◌◉◌◉◉◉◉◉◌◉◌◉◉◉◉◉◉◌◉◉◉◉◌◉◉
◌◉◌◌◌◌◌◌◌◌◌◌◌◌◌◌✻◌◉◌◌◌◌✻◌◉◉◉◉◌◌◌◌◌◌◌◌◌✻◌◉◌◉◉◌◉◌◌◌✻◌◉◌◉◌◌◌◌✻◌◉◌◌✻◌◉✻
◉◉◌◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◌◉◉◉◉◉◉◌◌✻◌◉◉◉◉◉◉◉◉◉◉◉◌◉✻◌◉◌◉◉◉◉◉◌◉◌◉◉◉◉◉◉◌◉◉◉◉◌
◉✻◌◉◌◌◌◌◌◌◌◌◌◌◌◌◌◌✻◌◉◌◌◌◌✻◌◉◉◉◉◌◌◌◌◌◌◌◌◌✻◌◉◌◉◉◌◉◌◌◌✻◌◉◌◉◌◌◌◌✻◌◉◌◌✻◌
◉◌◉◉◌◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◌◉◉◉◉◉◉◌◌✻◌◉◉◉◉◉◉◉◉◉◉◉◌◉✻◌◉◌◉◉◉◉◉◌◉◌◉◉◉◉◉◉◌◉◉◉
✻◌◉✻◌◉◌◌◌◌◌◌◌◌◌◌◌◌◌◌✻◌◉◌◌◌◌✻◌◉◉◉◉◌◌◌◌◌◌◌◌◌✻◌◉◌◉◉◌◉◌◌◌✻◌◉◌◉◌◌◌◌✻◌◉◌◌
◉◉◉◌◉◉◌◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◌◉◉◉◉◉◉◌◌✻◌◉◉◉◉◉◉◉◉◉◉◉◌◉✻◌◉◌◉◉◉◉◉◌◉◌◉◉◉◉◉◉◌◉
◌◌✻◌◉✻◌◉◌◌◌◌◌◌◌◌◌◌◌◌◌◌✻◌◉◌◌◌◌✻◌◉◉◉◉◌◌◌◌◌◌◌◌◌✻◌◉◌◉◉◌◉◌◌◌✻◌◉◌◉◌◌◌◌✻◌◉
◌◉◉◉◉◌◉◉◌◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◌◉◉◉◉◉◉◌◌✻◌◉◉◉◉◉◉◉◉◉◉◉◌◉✻◌◉◌◉◉◉◉◉◌◉◌◉◉◉◉◉◉
◌◉◌◌✻◌◉✻◌◉◌◌◌◌◌◌◌◌◌◌◌◌◌◌✻◌◉◌◌◌◌✻◌◉◉◉◉◌◌◌◌◌◌◌◌◌✻◌◉◌◉◉◌◉◌◌◌✻◌◉◌◉◌◌◌◌✻
◉◉◌◉◉◉◉◌◉◉◌◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◌◉◉◉◉◉◉◌◌✻◌◉◉◉◉◉◉◉◉◉◉◉◌◉✻◌◉◌◉◉◉◉◉◌◉◌◉◉◉◉
◌✻◌◉◌◌✻◌◉✻◌◉◌◌◌◌◌◌◌◌◌◌◌◌◌◌✻◌◉◌◌◌◌✻◌◉◉◉◉◌◌◌◌◌◌◌◌◌✻◌◉◌◉◉◌◉◌◌◌✻◌◉◌◉◌◌◌
◉◉◉◉◌◉◉◉◉◌◉◉◌◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◌◉◉◉◉◉◉◌◌✻◌◉◉◉◉◉◉◉◉◉◉◉◌◉✻◌◉◌◉◉◉◉◉◌◉◌◉◉
◌◌◌✻◌◉◌◌✻◌◉✻◌◉◌◌◌◌◌◌◌◌◌◌◌◌◌◌✻◌◉◌◌◌◌✻◌◉◉◉◉◌◌◌◌◌◌◌◌◌✻◌◉◌◉◉◌◉◌◌◌✻◌◉◌◉◌
◉◉◉◉◉◉◌◉◉◉◉◌◉◉◌◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◌◉◉◉◉◉◉◌◌✻◌◉◉◉◉◉◉◉◉◉◉◉◌◉✻◌◉◌◉◉◉◉◉◌◉◌
◉◌◌◌◌✻◌◉◌◌✻◌◉✻◌◉◌◌◌◌◌◌◌◌◌◌◌◌◌◌✻◌◉◌◌◌◌✻◌◉◉◉◉◌◌◌◌◌◌◌◌◌✻◌◉◌◉◉◌◉◌◌◌✻◌◉◌
◉◌◉◉◉◉◉◉◌◉◉◉◉◌◉◉◌◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◌◉◉◉◉◉◉◌◌✻◌◉◉◉◉◉◉◉◉◉◉◉◌◉✻◌◉◌◉◉◉◉◉◌
◉◌◉◌◌◌◌✻◌◉◌◌✻◌◉✻◌◉◌◌◌◌◌◌◌◌◌◌◌◌◌◌✻◌◉◌◌◌◌✻◌◉◉◉◉◌◌◌◌◌◌◌◌◌✻◌◉◌◉◉◌◉◌◌◌✻◌
◉◌◉◌◉◉◉◉◉◉◌◉◉◉◉◌◉◉◌◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◌◉◉◉◉◉◉◌◌✻◌◉◉◉◉◉◉◉◉◉◉◉◌◉✻◌◉◌◉◉◉◉
✻◌◉◌◉◌◌◌◌✻◌◉◌◌✻◌◉✻◌◉◌◌◌◌◌◌◌◌◌◌◌◌◌◌✻◌◉◌◌◌◌✻◌◉◉◉◉◌◌◌◌◌◌◌◌◌✻◌◉◌◉◉◌◉◌◌◌
◉◉◉◌◉◌◉◉◉◉◉◉◌◉◉◉◉◌◉◉◌◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◌◉◉◉◉◉◉◌◌✻◌◉◉◉◉◉◉◉◉◉◉◉◌◉✻◌◉◌◉◉
◌◌✻◌◉◌◉◌◌◌◌✻◌◉◌◌✻◌◉✻◌◉◌◌◌◌◌◌◌◌◌◌◌◌◌◌✻◌◉◌◌◌◌✻◌◉◉◉◉◌◌◌◌◌◌◌◌◌✻◌◉◌◉◉◌◉◌
◉◉◉◉◉◌◉◌◉◉◉◉◉◉◌◉◉◉◉◌◉◉◌◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◌◉◉◉◉◉◉◌◌✻◌◉◉◉◉◉◉◉◉◉◉◉◌◉✻◌◉◌
◉◌◌◌✻◌◉◌◉◌◌◌◌✻◌◉◌◌✻◌◉✻◌◉◌◌◌◌◌◌◌◌◌◌◌◌◌◌✻◌◉◌◌◌◌✻◌◉◉◉◉◌◌◌◌◌◌◌◌◌✻◌◉◌◉◉◌
◉◌◉◉◉◉◉◌◉◌◉◉◉◉◉◉◌◉◉◉◉◌◉◉◌◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◉◌◉◉◉◉◉◉◌◌✻◌◉◉◉◉◉◉◉◉◉◉◉◌◉✻◌
◉◌◉◌◌◌✻◌◉◌◉◌◌◌◌✻◌◉◌◌✻◌◉✻◌◉◌◌◌◌◌◌◌◌◌◌◌◌◌◌✻◌◉◌◌◌◌✻◌◉◉◉◉◌◌◌◌◌◌◌◌◌✻◌◉◌◉
```


CHANGELOG.md
# Changelog

### [4.1] - 2021-08-22

- Add custom tests

### [4.0] - 2021-08-13

- Added second order + some refactos

### [3.3] - 2021-05-04

- Single page to generate all states, loaded in javascript

### [3.2] - 2021-04-23

- Added a text renderer
- Optimize the defaults parameters management
- Cleaned the computation in `CellularAutomata`

## [3.2] - 2021-04-23

- Added a text renderer
- Optimize the defaults parameters management
- Cleaned the computation in `CellularAutomata`

## [3.1] - 2021-03-10

- Finished refacto: finally following SOLID principles
- Now supporting any number of states (>2)

## [3.0] - 2021-03-10

- Another refactoring

## [2.0] - 2016-10-20

- Class refactoring

## [1.1] - 2016-30-30

- Added presentation page

## [1.0] - 2016-30-20

- Updated project.


lib/Lib.php
<?php

/**
 * Class Lib.
 */
class Lib
{
    // Defaults
    const COLOR = '#000000';
    const COLOR_BG = '#ffffff';
    const COLOR_2 = '#aacc00';
    const COLOR_3 = '#0055cc';

    /**
     * Get parameters from $_GET.
     */
    public static function getColor()
    {
        return $_GET['color'] ?? self::COLOR;
    }

    public static function getBgColor()
    {
        return $_GET['bgcolor'] ?? self::COLOR_BG;
    }

    public static function getColor2()
    {
        return $_GET['color2'] ?? self::COLOR_2;
    }

    public static function getColor3()
    {
        return $_GET['color3'] ?? self::COLOR_3;
    }

    public static function getRandomStart()
    {
        if (!array_key_exists('randomstart', $_GET)) {
            return false;
        }

        return 1 == $_GET['randomstart'] || 'on' == $_GET['randomstart'];
    }

    public static function getPixelSize()
    {
        return $_GET['pixel'] ?? 1;
    }

    public static function getWidth()
    {
        return $_GET['width'] ?? 500;
    }

    public static function getHeight()
    {
        return $_GET['height'] ?? 500;
    }

    public static function getRandomRule()
    {
        if (!array_key_exists('random', $_GET)) {
            return false;
        }

        return 1 == $_GET['random'] || 'on' == $_GET['randomstart'];
    }

    public static function getRuleNumber()
    {
        return $_GET['rule'] ?? null;
    }

    /**
     * Returns a rule number, from the paramater or randomly.
     *
     * @param {int} the number of states
     * @param mixed $states
     */
    public function whichRule($states): int
    {
        $maxRule = $this->computeMaxRule($states);
        //self::setMaxRules($states);

        
        if (false != self::getRandomRule() || 0 == self::getRuleNumber()) {
            return rand(0, $maxRule);
        }
        $rule = intval(self::getRuleNumber());
        if ($rule > $maxRule) {
            throw new Exception(sprintf(
                'Cannot use rule #%d (max: %d)',
                $rule,
                $maxRule
            ));
        }

        return $rule;
    }

    /**
     * Returns the dimension of the final image, and the length of a pixel.
     *
     * @return array column, generationsNb, pixelLength
     */
    public static function getSize(): array
    {
        $columns = self::getWidth();
        $generationsNb = self::getHeight();

        if (intval(self::getPixelSize()) > 1) {
            $pixelLength = intval(self::getPixelSize());
            $columns *= $pixelLength;
            $generationsNb *= $pixelLength;
        } else {
            $pixelLength = 1;
        }

        return [
            'columns' => $columns,
            'generationsNb' => $generationsNb,
            'pixelLength' => $pixelLength,
        ];
    }

    /**
     * The index, when in bin[/trin]ary, will represent the state of the three current cells,
     *  and the number (0/1[/2]), will represent the state of the resulting cell.
     *
     * @param {int} $ruleNumber
     * @param {int} $states     digital base
     *
     * @return array an "associative" array corresponding to the rule number
     */
    public static function ruleToArray($ruleNumber, $states): array
    {
        $toBaseN = sprintf('%08d', intval(base_convert($ruleNumber, 10, $states)));

        return array_reverse(str_split(strval($toBaseN)));
    }

    /**
     * Decode Color from string, either rrr,vvv,bbb or #rrvvbb
     * and allocate this color to the Image resource.
     *
     * @param {string}   $color    color in hexa or decimal format
     * @param {resource} $theImage the image resource
     *
     * @throws Exception if the format could not be understood
     *
     * @return int (ImageColorAllocate)
     */
    public static function decodeColor($color, $theImage): int
    {
        preg_match('/#([0-9a-fA-F]+)/i', $color, $match_hexa);
        preg_match('/\d+,\d+,\d+/i', $color, $match_rgb);
        if (count($match_hexa) > 0) { // #CCDD99
            $hexaNbs = str_split($match_hexa[1], 2);

            return imagecolorallocate(
                $theImage,
                intval(base_convert($hexaNbs[0], 16, 10)),
                intval(base_convert($hexaNbs[1], 16, 10)),
                intval(base_convert($hexaNbs[2], 16, 10))
            );
        }
        if (count($match_rgb) > 0) {  // rgb(200,220,180)
            $colorNbs = explode(',', $color);

            return imagecolorallocate(
                $theImage,
                intval($colorNbs[0]),
                intval($colorNbs[1]),
                intval($colorNbs[2])
            );
        }

        throw new Exception('The color '.$color.' could not be decoded.');
    }

    protected function computeMaxRule($states)
    {
        // TODO find the correct formula...?
        switch ($states) {
            case 2:
                return 256; // 2 ** $states ** 3
            case 3:
                return 52486;
            case 4:
                return 500000; // approx. 274875000000;
            default:
                throw new Exception(sprintf(
                    'Cannot process other states than 2, 3, 4 (%s (%s) given)',
                    $states,
                    gettype($states)
                ));
        }
    }
}
